fecgov/openFEC
# Add historical data of candidate_inactve to ofec_candidate_detail_mv

## Summary 

- Resolves [#3810](https://github.com/fecgov/openFEC/issues/3810)

- New column inactive_election_years has been added to ofec_candidate_detail_mv, and it hold an list of inactive election years each candidate had.  For example, 
S00000000,  inactive_election_years=[2004,2016], election_years=[1998,2004,2010,2016,2022]
- When election_year is not passed, the query would retain its current behavior by comparing with the old boolean candidate_inactive column.  By default, the API returns candidates and their data as of latest election_year without election_year.  It is as of latest FEC cycle as well .  The reason of acting this way is ofec_candidate_detail_mv takes the latest information per candidate.
- When election_year is passed to endpoint, the query would compare with the new array column to decide if candidates are active or inactive.  The logic is as this:
if is_active_candidate = true, 
    election_year = 2022:      S00000000 is returned
    election_year = 2016:       S00000000 is not returned
    election_year =[2016,2022]:   S00000000 is returned 
if is_active_candidate = false, 
    election_year = 2016:       S00000000 is returned
    election_year =[2016,2022]:   S00000000 is returned
    election_year = 2022:  S00000000 is not returned
    


## How to test the changes locally

- download feature/3810-ofec-candidate-detail-mv.
- run migration file V0141 (or run invoke create_sample_db) to make sure migration files runs without errors
- To test API:  candidate: /candidates/
   set local API to DEV database, and 
   In models/candidates.py file, make the following changes: 

> class BaseConcreteCandidate:
     tablename = 'ofec_candidate_detail_mv_tmp_hc'

    class Candidate:
      tablename = 'ofec_candidate_detail_mv_tmp_hc'

    class CandidateDetail:
     tablename = 'ofec_candidate_detail_mv_tmp_hc'
    
In models/links.py file, make the following changes:

> class CandidateCommitteeLink:
    candidate_id = db.Column('cand_id', db.String, db.ForeignKey('ofec_candidate_detail_mv_tmp_hc.candidate_id'))

Run this API and it would return S6AZ00019 as an active candidate for 2016
http://127.0.0.1:5000/v1/candidates/?sort_nulls_last=false&election_year=2016&is_active_candidate=true&party=REP&sort_null_only=false&per_page=20&sort_hide_null=false&office=S&sort=name&page=1&state=AZ
    

Run this API, and S6AZ00019 is listed as inactive candidate for 2022
http://127.0.0.1:5000/v1/candidates/?sort_nulls_last=false&election_year=2022&is_active_candidate=false&party=REP&sort_null_only=false&per_page=20&sort_hide_null=false&office=S&sort=name&page=1&state=AZ
